### PR TITLE
add expressions and payload transformations to commands

### PR DESCRIPTION
**Pull request by AlvaroVega**
Issues: https://github.com/telefonicaid/iotagent-json/issues/634 https://github.com/telefonicaid/iotagent-json/issues/627

related and needs https://github.com/telefonicaid/iotagent-node-lib/pull/1209

The possible payload transformations are: 

 https://github.com/telefonicaid/iotagent-ul/blob/243b2807aad41880ec911fd8ca8fe2e19b5cbfb2/lib/ulParser.js#L243-L257

- [ ] Update doc

### CODE REVIEW

File: docs/usermanual.md [hunk 1]
```diff
@@ -105,6 +105,37 @@ a transformation expression in the IoTAgent). In this cases, when a reverse expr
...
+}
+```
+
+Additionally a command could define a `payloadType` in their definnition with the aim to transform payload command with
+the following meanings:
+
+-   **binaryfromstring**: Payload will transformed into a be Buffer after read it form a string.
+-   **binaryfromhex**: Payload will transformed into a be Buffer after read it form a string hex.
+-   **binaryfromjson**: Payload will transformed into a be Buffer after read it form a JSON string.
+-   **json**: Paylaod will be stringify form a JSON.
```
Comment (fgalan) [inline docs/usermanual.md:136]:
```suggestion
-   **json**: Payload will be stringify form a JSON.
```

File: lib/ulParser.js [hunk 2]
```diff
@@ -239,13 +273,27 @@ function result(payload) {
...
 function createCommandPayload(device, command, attributes) {
     function addAttributes(current, key) {
-        return current + '|' + key + '=' + attributes[key];
+        //let cmd = device && device.commands.find((att) => att.name === key);
+        let value = attributes[key];
+        // if (cmd && cmd.expression) {
+        //     let parser = iotAgentLib.dataPlugins.expressionTransformation;
+        //     let ctxt = device.staticAttributes.concat({key: value});
+        //     value = parser.applyExpression(cmd.expression, ctxt, device);
+        // }
```
Comment (fgalan) [inline lib/ulParser.js:282]: Leftover comments?

File: docs/usermanual.md (hunk 1 above)
Comment (fgalan) [reply docs/usermanual.md:136]: form -> from? (unsure)

File: docs/usermanual.md [hunk 3]
```diff
@@ -105,6 +105,37 @@ a transformation expression in the IoTAgent). In this cases, when a reverse expr
...
+```
+
+and when command will be executed the command value will be the result of apply value to defined expression. Followig
+the example case the command will be:
+
+```json
+{
+    "set": 0
+}
+```
```
Comment (fgalan) [inline docs/usermanual.md:128]: Should be in UL format? Something like `set|0` (unsure)

File: lib/ulParser.js (hunk 2 above)
Comment (AlvaroVega) [reply lib/ulParser.js]: fixed in https://github.com/telefonicaid/iotagent-ul/pull/550/commits/9908cba261d7f9b253df8a65010adba27ac273e2

File: CHANGES_NEXT_RELEASE [hunk 4]
```diff
@@ -1,3 +1,4 @@
+Add: apply expression and payload transformations to commands (iota-json#634, iota-json#627
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:1]:
```suggestion
Add: apply expression and payload transformations to commands (iota-json#634, iota-json#627)
```

**Review by fgalan — APPROVED** [review]
LGTM (but waiting to merge https://github.com/telefonicaid/iotagent-node-lib/pull/1209 before merging this one)

**Pull request merged**